modules/Movecraft/src/main/java/net/countercraft/movecraft/async/rotation/RotationTask.java

```diff
@@ -195,6 +207,65 @@ protected void execute() {
             }
             return;
         }
+        boolean moveDown = true;
+
+        for (MovecraftLocation ml : newHitBox){
+            if (ml.getY() > newHitBox.getMinY()){
+                continue;
+            }
+            Material typeUnder = ml.translate(0, -1, 0).toBukkit(w).getBlock().getType();
+
+            if (typeUnder != Material.AIR && !craft.getType().getPassthroughBlocks().contains(typeUnder)){
+                moveDown = false;
+            }
+        }
+        if (craft.getType().getCanHover() && craft.getType().getHoverLimit() > 0){
+            MovecraftLocation midPoint = newHitBox.getMidPoint();
+            int centreMinY = newHitBox.getLocalMinY(midPoint.getX(), midPoint.getZ());
+            int groundY = centreMinY;
+            while (w.getBlockAt(midPoint.getX(), groundY - 1, midPoint.getZ()).getType() == Material.AIR || craft.getType().getPassthroughBlocks().contains(w.getBlockAt(midPoint.getX(), groundY - 1, midPoint.getZ()).getType())){
+                groundY--;
+            }
+            if (centreMinY - groundY <= craft.getType().getHoverLimit() && moveDown){
+                moveDown = false;
+            }
+        }
+        //Don't process gravity on the second movement processing
+        if (craft.getLastGravityOnRotateTime() + 1e9 > System.nanoTime()){
+            moveDown = false;
+            moveUp = false;
+        }
+        processingGravity = moveDown || moveUp;
+        if (moveUp){
+            if (!craft.isNotProcessing()){
+                craft.setProcessing(false);
+            }
+            craft.setLastGravityOnRotateTime(System.nanoTime());
+            craft.translate(0, 1, 0);
+            new BukkitRunnable(){
+
+                @Override
+                public void run() {
+                    craft.rotate(rotation, originPoint, false);
+                }
+            }.runTaskLaterAsynchronously(Movecraft.getInstance(), 2);
+            return;
+        }
+        else if (moveDown){
+            if (!craft.isNotProcessing()){
+                craft.setProcessing(false);
+            }
+            craft.setLastGravityOnRotateTime(System.nanoTime());
+            craft.rotate(rotation, originPoint, false);
+            new BukkitRunnable(){
+                @Override
+                public void run() {
+                    craft.translate(0, -1, 0);
+
+                }
+            }.runTaskLaterAsynchronously(Movecraft.getInstance(), 2);
+            return;
+        }
```

Comment:
The only thing different between what is processed in moveup and movedown is the vector at which the craft translates. There's no need to repeat all of the code twice.

Reply:
The order of when rotate and translate is called is also different